README.md

# IRAN BANKS 🌍💫

By using this package, you will be able to easily create a beautiful bottom navigation bar in flutter and use it on diffrent Os, here is the documantation about how to use it💡

🟢داکیومنت فارسی در مورد نحوه استفاده از این پکیج در آینده نوشته خواهد شد


![IranBanks](https://github.com/TahaAhvazi/iranbanks/assets/86960501/055c85ae-bcc3-4817-91b7-92d9f4725a7a)



![image](https://github.com/TahaAhvazi/iranbanks/assets/86960501/663fab33-cad1-4cc0-a058-e7902c4447f7)



## How to use⁉️

#### Use it wherever you want: ✅

```
IranBanks(
height: 200,
width: 300,
// All this field should be String
usernameAndFamily: "طاها اهوازی",
expCartMonth: "06",
// All this field should be String
expCartYear: "05",
// All this field should be String
shabbaNumber: "23456756736754747846",
// All this field should be String
cvv2: "6118",
//Cart number is int
cartNumber: 123456789101112,
// All this field should be String
shomarehHesab: "12344567845234",
//Use the default bank type from enums
banktype: bankType,
),
```

The arguments needed for (ClassicButtomNavBar) 🟢

| Parameter | Type     | Description                |
| :-------- | :------- | :------------------------- |
| `usernameAndFamily` | `String` | **Required**. color of container |


| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `expCartMonth`      | `String` | **Required**. data of exp Cart in month |

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `expCartYear`      | `double` |  **Required**. data of exp Cart in year|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `height`      | `double` | **Required**. height of the Iran banks|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `width`      | `double` | **Required**. width of the Iran banks : |


| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `shabbaNumber`      | `String` | **required**. shabbaNumber of the user|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `cvv2`      | `String` | **required**. cvv2 of the user|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `cartNumber`      | `int` | **required**. cartNumber of the user|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `shomarehHesab`      | `String` | **required**. shomarehHesab of the user|

| Parameter | Type     | Description                       |
| :-------- | :------- | :-------------------------------- |
| `banktype`      | `BankTypes` | **required**. banktype of the enum|


## Author👦

- [@Taha ahvazi ](https://github.com/TahaAhvazi)


## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
## 0.0.1
## 0.0.2 
* UPDATE: MAke it online
## 0.0.3
* BUG: Fix Some bugs


lib/src/iran_banks.dart
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

enum BankType {
  meli,
  ayandeh,
  sepah,
  keshavarzi,
  maskan,
  sanatvamadan,
  saderat,
  taavon,
  postbank,
  refah,
  melat,
}

class IranBanks extends StatefulWidget {
  const IranBanks({
    Key? key,
    required this.cartNumber,
    required this.height,
    required this.width,
    required this.banktype,
    required this.usernameAndFamily,
    required this.expCartMonth,
    required this.expCartYear,
    required this.shabbaNumber,
    required this.cvv2,
    required this.shomarehHesab,
    this.heighBackImage = 150,
    this.widthBackImage = 300,
    this.fontFamily = '',
  }) : super(key: key);
  final int cartNumber;
  final String expCartMonth;
  final String expCartYear;
  final String shabbaNumber;
  final String usernameAndFamily;
  final String cvv2;
  final String shomarehHesab;
  final double height, width;
  final double heighBackImage;
  final double widthBackImage;
  final String fontFamily;
  final BankType banktype;
  @override
  // ignore: library_private_types_in_public_api
  _IranBanksState createState() => _IranBanksState();
}

class _IranBanksState extends State<IranBanks> {
  late bool _showFrontSide;
  late bool _flipXAxis;
  int count = 0;

  String setIMageAdress() {
    String imagePath = ''; // Initialize imagePath

    // Determine the image path based on the bankType
    switch (widget.banktype) {
      case BankType.meli:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/melilogo_scqr.png';
        return imagePath;
      //add address of images network on the host to recive them
      case BankType.ayandeh:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/ayandehlogo_u8rm.png';
        return imagePath;
      case BankType.sepah:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/sepahlogo_rzc6.png';
        return imagePath;
      case BankType.keshavarzi:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/keshavarzilogo_enos.png';
        return imagePath;
      case BankType.maskan:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/maskanlogo_gg5f.png';
        return imagePath;
      case BankType.sanatvamadan:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/sanatvamadanlogo_8ecj.png';
        return imagePath;
      case BankType.saderat:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/saderatlogo_ntvq.png';
        return imagePath;
      case BankType.taavon:
        imagePath = 'https://s6.uupload.ir/files/taavonlogo_o3bi.png';
        return imagePath;
      case BankType.postbank:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/postbanklogo_j5rs.png';
        return imagePath;
      case BankType.refah:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/refahlogo_h5n1.png';
        return imagePath;
      case BankType.melat:
        //add address of images network on the host to recive them
        imagePath = 'https://s6.uupload.ir/files/melatlogo_mzza.png';
        return imagePath;
    }
  }

  Color setColorBank() {
    // Determine the Color based on the bankType
    switch (widget.banktype) {
      case BankType.meli:
        return Colors.blue;
      case BankType.ayandeh:
        //set colors for each bank base on its icon
        return const Color.fromARGB(255, 187, 106, 0);

      case BankType.sepah:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 185, 224, 255);

      case BankType.keshavarzi:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 243, 240, 33);

      case BankType.maskan:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 224, 91, 1);

      case BankType.sanatvamadan:
        //set colors for each bank base on its icon

        return const Color.fromARGB(129, 0, 0, 0);

      case BankType.saderat:
        //set colors for each bank base on its icon

        return const Color.fromARGB(174, 0, 17, 253);

      case BankType.taavon:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 88, 173, 241);

      case BankType.postbank:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 2, 146, 9);

      case BankType.refah:
        //set colors for each bank base on its icon

        return const Color.fromARGB(255, 107, 33, 243);

      case BankType.melat:
        //set colors for each bank base on its icon

        return Colors.red;
    }
  }

  @override
  void initState() {
    super.initState();
    _showFrontSide = true;
    _flipXAxis = true;
  }

  @override
  Widget build(BuildContext context) {
    return buildFlipAnimation();
  }

  void _changeRotationAxis(bool isX) {
    setState(() {
      isX ? _flipXAxis = true : _flipXAxis = false;
    });
  }

  void _switchCard() {
    setState(() {
      _showFrontSide = !_showFrontSide;
    });
  }

  Widget buildFlipAnimation() {
    return GestureDetector(
      onHorizontalDragUpdate: (details) {
        if (details.delta.dx > 0) {
          setState(() {
            if (kDebugMode) {
              print("RIGHT ROTATION");
            }
            if (count == 0) {
              _changeRotationAxis(true);
              _switchCard();
              count++;
            }
            if (count != 0) {
              Future.delayed(const Duration(milliseconds: 600))
                  .then((value) => count = 0);
            }
          });
        }
        if (details.delta.dx < 0) {
          setState(() {
            if (kDebugMode) {
              print("LEFT ROTATION");
            }
            if (count == 0) {
              _changeRotationAxis(true);
              _switchCard();
              count++;
            }
            if (count == 0) {
              _changeRotationAxis(true);
              _switchCard();
              count++;
            }
            if (count != 0) {
              Future.delayed(const Duration(milliseconds: 500))
                  .then((value) => count = 0);
            }
          });
        }
      },
      onVerticalDragUpdate: (details) {
        if (details.delta.dy > 0) {
          setState(() {
            if (count == 0) {
              _changeRotationAxis(false);
              _switchCard();
              count++;
            }
            if (count == 0) {
              _changeRotationAxis(false);
              _switchCard();
              count++;
            }
            if (count != 0) {
              Future.delayed(const Duration(milliseconds: 500))
                  .then((value) => count = 0);
            }
          });
          if (kDebugMode) {
            print("DOWN ROTATION");
          }
        }
        if (details.delta.dy < 0) {
          setState(() {
            if (count == 0) {
              _changeRotationAxis(false);
              _switchCard();
              count++;
            }
            if (count == 0) {
              _changeRotationAxis(false);
              _switchCard();
              count++;
            }
            if (count != 0) {
              Future.delayed(const Duration(milliseconds: 500))
                  .then((value) => count = 0);
            }
          });
          if (kDebugMode) {
            print("UP ROTATION");
          }
        }
      },
      child: AnimatedSwitcher(
        duration: const Duration(milliseconds: 800),
        transitionBuilder: __transitionBuilder,
        layoutBuilder: (widget, list) => Stack(children: [widget!, ...list]),
        switchInCurve: Curves.easeInBack,
        switchOutCurve: Curves.easeInBack.flipped,
        child: _showFrontSide ? _buildFront() : _buildRear(),
      ),
    );
  }

  Widget __transitionBuilder(Widget widget, Animation<double> animation) {
    final rotateAnim = Tween(begin: pi, end: 0.0).animate(animation);
    return AnimatedBuilder(
      animation: rotateAnim,
      child: widget,
      builder: (context, widget) {
        final isUnder = (ValueKey(_showFrontSide) != widget?.key);
        var tilt = ((animation.value - 0.5).abs() - 0.5) * 0.003;
        tilt *= isUnder ? -1.0 : 1.0;
        final value =
            isUnder ? min(rotateAnim.value, pi / 2) : rotateAnim.value;
        return Transform(
          transform: _flipXAxis
              ? (Matrix4.rotationY(value)..setEntry(3, 1, tilt))
              : (Matrix4.rotationX(value)..setEntry(3, 2, tilt)),
          alignment: Alignment.center,
          child: widget,
        );
      },
    );
  }

  Widget _buildFront() {
    return __buildLayout(
      key: const ValueKey(true),
      backgroundColor: const Color.fromARGB(255, 252, 252, 252),
      cartNumber: widget.cartNumber,
      height: widget.height,
      width: widget.width,
      faceName: "Front",
      child: Stack(
        children: [
          Positioned(
            bottom: 0,
            left: 0,
            right: 0,
            child: ClipRRect(
              borderRadius: const BorderRadius.all(Radius.circular(8)),
              child: ColorFiltered(
                colorFilter: ColorFilter.mode(setColorBank(), BlendMode.color),
                child: Image.network(
                  "https://s6.uupload.ir/files/back_c3q4.png",
                  height: widget.heighBackImage,
                  width: widget.heighBackImage,
                  fit: BoxFit.fill,
                ),
              ),
            ),
          ),
          Positioned(
            left: 0,
            right: 0,
            bottom: 0,
            top: 0,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Image.network(
                      "https://s6.uupload.ir/files/shetab_ftnh.png",
                      height: 50,
                      width: 50,
                    ),
                    Image.network(
                      setIMageAdress(),
                      height: 50,
                      width: 50,
                    ),
                  ],
                ),
                Padding(
                  padding: const EdgeInsets.only(top: 10.0),
                  child: Text(
                    "IR${widget.shabbaNumber}",
                    style: const TextStyle(
                      color: Colors.white,
                      fontSize: 16,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                ),
                const SizedBox(
                  height: 10,
                ),
                Container(
                  height: 65,
                  width: 250,
                  decoration: BoxDecoration(
                    color: const Color.fromARGB(204, 255, 255, 255),
                    borderRadius: BorderRadius.circular(8.0),
                  ),
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Text(
                        widget.cartNumber.toString(),
                        style: TextStyle(
                          fontFamily: widget.fontFamily,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(left: 8.0, right: 8.0),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            Text(
                              "${widget.expCartYear}/${widget.expCartMonth}",
                              style: TextStyle(
                                fontFamily: widget.fontFamily,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                            Text(
                              widget.usernameAndFamily,
                              style: TextStyle(
                                fontFamily: widget.fontFamily,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                          ],
                        ),
                      )
                    ],
                  ),
                ),
                const SizedBox(
                  height: 5,
                ),
                Container(
                  height: 30,
                  width: 250,
                  decoration: BoxDecoration(
                    color: const Color.fromARGB(204, 255, 255, 255),
                    borderRadius: BorderRadius.circular(8.0),
                  ),
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Padding(
                        padding: const EdgeInsets.only(left: 8.0, right: 8.0),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            Text(
                              "CVV2:${widget.cvv2}",
                              style: TextStyle(
                                fontFamily: widget.fontFamily,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                            Text(
                              widget.shomarehHesab,
                              style: TextStyle(
                                fontFamily: widget.fontFamily,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                          ],
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildRear() {
    return __buildLayout(
        key: const ValueKey(false),
        width: widget.width,
        height: widget.height,
        cartNumber: widget.cartNumber,
        backgroundColor: setColorBank(),
        faceName: "Rear",
        child: Column(
          children: [
            const SizedBox(
              height: 10,
            ),
            Container(
              width: 300,
              height: 50,
              decoration: const BoxDecoration(
                color: Colors.black,
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(8),
                  topRight: Radius.circular(8),
                ),
              ),
            ),
            const SizedBox(
              height: 7,
            ),
            Container(
              width: 280,
              height: 50,
              decoration: const BoxDecoration(
                color: Colors.white,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: [
                  Text(
                    "استفاده از این کارت صرفا مخصوص دارنده کارت میباشد",
                    style: TextStyle(
                      fontFamily: widget.fontFamily,
                      fontSize: 12,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  Text(
                    "در صورتی که صاحب کارت بانکی نیستید با بانک در میان بگذارید",
                    style: TextStyle(
                      fontFamily: widget.fontFamily,
                      fontSize: 12,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                ],
              ),
            ),
          ],
        ));
  }

  Widget __buildLayout(
      {required Key key,
      required Widget child,
      required String faceName,
      required double height,
      required double width,
      int? cartNumber,
      required Color backgroundColor}) {
    return Container(
      key: key,
      height: height,
      width: width,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(20.0),
        color: backgroundColor,
      ),
      child: child,
    );
  }
}


example/lib/main.dart
import 'package:example/package.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const Test());
}

class Test extends StatefulWidget {
  const Test({super.key});

  @override
  State<Test> createState() => _TestState();
}

class _TestState extends State<Test> {
  //Create a default instance of enum for our initial Bank type
  BankType bankType = BankType.melat;
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        backgroundColor: const Color.fromARGB(255, 255, 255, 255),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              // call the package class name to start using the iran bank cart
              IranBanks(
                height: 200,
                width: 300,
                // All this field should be String
                usernameAndFamily: "طاها اهوازی",
                expCartMonth: "06",
                // All this field should be String
                expCartYear: "05",
                // All this field should be String
                shabbaNumber: "23456756736754747846",
                // All this field should be String
                cvv2: "6118",
                //Cart number is int
                cartNumber: 123456789101112,
                // All this field should be String
                shomarehHesab: "12344567845234",
                //Use the default bank type from enums
                banktype: bankType,
              ),
              const SizedBox(
                height: 20,
              ),
              const SizedBox(height: 9),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.meli;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/melilogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.ayandeh;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/ayandehlogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.sepah;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/sepahlogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.keshavarzi;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/keshavarzilogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.maskan;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/maskanlogo.png"),
                    ),
                  ),
                ],
              ),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                children: [
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.sanatvamadan;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/sanatvamadanlogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.saderat;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/saderatlogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.taavon;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/taavonlogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.postbank;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/postbanklogo.png"),
                    ),
                  ),
                  GestureDetector(
                    onTap: () {
                      setState(() {
                        bankType = BankType.refah;
                      });
                    },
                    child: Container(
                      height: 70,
                      width: 70,
                      color: Colors.white,
                      child: Image.asset("assets/refahlogo.png"),
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
